Library/Homebrew/cmd/update.sh

```diff
@@ -334,9 +334,26 @@ EOS
     UPSTREAM_BRANCH="$(upstream_branch)"
     # the refspec ensures that the default upstream branch gets updated
     (
-      git fetch "${QUIET_ARGS[@]}" origin \
-        "refs/heads/$UPSTREAM_BRANCH:refs/remotes/origin/$UPSTREAM_BRANCH" || \
-          odie "Fetching $DIR failed!"
+      UPSTREAM_REPOSITORY_URL="$(git config remote.origin.url)"
+      if [[ "$UPSTREAM_REPOSITORY_URL" = "https://github.com/"* ]]
+      then
+        UPSTREAM_REPOSITORY="$(echo "$UPSTREAM_REPOSITORY_URL" | \
+          sed -e "s|^https://github.com/||" -e "s|.git$||")"
+        UPSTREAM_BRANCH_LOCAL_SHA="$(git rev-parse "refs/remotes/origin/$UPSTREAM_BRANCH")"
+        # Only try to `git fetch` when the upstream branch is at a different SHA
+        # (so the API does not return 304: unmodified).
+        UPSTREAM_SHA_HTTP_CODE="$(curl --silent '--max-time' 3 \
+           --output /dev/null --write-out "%{http_code}" \
+           -H "Accept: application/vnd.github.chitauri-preview+sha" \
+           -H "If-None-Match: \"$UPSTREAM_BRANCH_LOCAL_SHA\"" \
+           "https://api.github.com/repos/$UPSTREAM_REPOSITORY/commits/$UPSTREAM_BRANCH")"
+      fi
```

Comment:
It could improve the code style by early return here. i.e

```bash
 if [[ "$UPSTREAM_REPOSITORY_URL" = "https://github.com/"* ]]
then
  ...
  [[ "$UPSTREAM_SHA_HTTP_CODE" = "304" ]] && return
fi
```